### PR TITLE
feat(ts/components/GroupedAutocomplete): introduce autocomplete component

### PR DESCRIPTION
**Pull request by firestack**
This creates a new control component for displaying autocomplete results and navigating that control with keyboard & mouse and moving and listening to the document focus state.

The component is then extended to use a hook to retrieve it's results from the back-end.

The the extended component is then embedded into the search form.

---

Asana Ticket: https://app.asana.com/0/1200180014510248/1204956501632027/f

### CODE REVIEW

File: assets/css/_search_form.scss
Comment (lemald) [inline assets/css/_search_form.scss:132]: **most nitpick-y of nitpicks (non-blocking):** isn't this actually squaring off, rather than rounding out, the bottom of the input bar?

Comment (firestack) [reply assets/css/_search_form.scss]: 🤣, thank you for catching this!!!
73bdffb34b3304656a02595cafd307433828c362

Comment (firestack) [reply assets/css/_search_form.scss]: another nit: should we move the border radius styling of the autocomplete into this file and make `.c-autocomplete`'s `border-radius` `4px` all around?

File: assets/src/components/groupedAutocomplete.tsx
Comment (firestack) [reply assets/src/components/groupedAutocomplete.tsx:554]: there's no reason it can't be, I was just leaving it open in case we ended up needing to add more keys. I think at the time I thought it looked cleaner, but we can update it.

Comment (lemald) [reply assets/src/components/groupedAutocomplete.tsx]: Makes sense, I guess it looked slightly weird to me but I have no strong preference in favor of changing it either (although I see you just did so in e5bed05).

File: assets/src/util/math.ts
Comment (lemald) [inline assets/src/util/math.ts:9]: **suggestion:** would be good to also add a test

File: assets/src/components/groupedAutocomplete.tsx
Comment (KaylaBrady) [inline assets/src/components/groupedAutocomplete.tsx:241]: question: When could the answer not be correct?

File: assets/src/util/operatorFormatting.tsx
Comment (lemald) [inline assets/src/util/operatorFormatting.tsx:4]: **suggestion (non-blocking):** not strictly in scope for this PR, but I think this could also be used to get rid of some existing logic that does roughly the same thing in `propertiesList.tsx`.

Comment (lemald) [reply assets/src/util/operatorFormatting.tsx:4]: Also, it would be good to have tests for this new module as well.

Comment (firestack) [reply assets/src/util/operatorFormatting.tsx:4]: I wanted to do so, and I avoided it since this was getting big, I'll open a separate PR (or open a ticket otherwise).

File: assets/src/components/groupedAutocomplete.tsx
Comment (KaylaBrady) [inline assets/src/components/groupedAutocomplete.tsx:365]: question (non-blocking): Is this needed as part of the props doc? It seems to me that this is taken care of within the component as long as a string value is passed for "Heading"

File: assets/src/util/math.ts
Comment (firestack) [reply assets/src/util/math.ts:9]: I thought there was a [polyfill](https://www.npmjs.com/package/ecma-proposal-math-extensions) for this I could use but it seems like the repo for the polyfill is gone. 

I had thought I was going to use this function a lot more, hence pulling it out into it's own function, do you feel strongly on keeping this and adding tests vs getting rid of it and inlining?

File: assets/src/components/groupedAutocomplete.tsx
Comment (KaylaBrady) [inline assets/src/components/groupedAutocomplete.tsx:566]: > Composite controls should have one tab stop

question: What is this referencing?

File: assets/src/components/groupedAutocomplete.tsx
Comment (KaylaBrady) [inline assets/src/components/groupedAutocomplete.tsx:708]: question: Should this filtering happen before slicing?

File: assets/src/components/groupedAutocomplete.tsx
Comment (firestack) [reply assets/src/components/groupedAutocomplete.tsx:365]: I had this [article](https://dev.to/s_aitchison/psa-stop-hard-coding-heading-levels-in-your-react-components-2ekp) in mind when writing this, in the case that we did end up with "automatic" heading levels in the app, I was thinking it'd be good to callout what interface was required. Though I agree it's not needed given it's simply trying to predict the future.

File: assets/src/util/math.ts
Comment (lemald) [reply assets/src/util/math.ts:9]: No real preference one way or another.

File: assets/src/util/operatorFormatting.tsx
Comment (firestack) [reply assets/src/util/operatorFormatting.tsx:10]: b215d090a4c82ac5219054f6a2c56a898f330e65

File: assets/src/components/groupedAutocomplete.tsx
Comment (firestack) [reply assets/src/components/groupedAutocomplete.tsx:241]: Uhhh, I don't have an answer for that! It's likely that it's always correct, I just didn't trust it at the time of writing.

File: assets/src/util/operatorFormatting.tsx
Comment (lemald) [inline assets/src/util/operatorFormatting.tsx:4]: **nitpick:** as a `const` I think this should follow our usual capitalization conventions (`defaultFallbackString`).

File: assets/src/util/operatorFormatting.tsx
Comment (lemald) [inline assets/src/util/operatorFormatting.tsx:14]: **nitpick:** we usually use underscore-prefixed variables to denote arguments that aren't read (as per Elixir convention), but `_options` is referenced in the function body.

File: assets/src/components/groupedAutocomplete.tsx
Comment (firestack) [reply assets/src/components/groupedAutocomplete.tsx:707]: a5b9495a6bf2bc0910e4a1633fbd70fce141f283

File: assets/src/util/operatorFormatting.tsx
Comment (firestack) [reply assets/src/util/operatorFormatting.tsx]: 9c174e5575e4c5061d5be97bed41fc5ea26dfbfe

File: assets/src/util/operatorFormatting.tsx
Comment (firestack) [reply assets/src/util/operatorFormatting.tsx]: I missed this while ensuring tests worked, thank you! 🙇🏻‍♀️

File: assets/src/util/math.ts
Comment (firestack) [reply assets/src/util/math.ts:9]: a55dbf86bd6ba9b874d26258a2354cf67917a592

File: assets/src/components/groupedAutocomplete.tsx
Comment (firestack) [reply assets/src/components/groupedAutocomplete.tsx:566]: I was certain this was in the WCAG or something but I must've unintentionally paraphrased where I read this from because I can't find it. Will revise.

Comment (firestack) [reply assets/src/components/groupedAutocomplete.tsx:566]: I thinking of removing this altogether and let arrow keys be the primary method of entering the control, because [WebAIM's documentation on Keyboard Accessibility](https://webaim.org/techniques/keyboard/) says that 

> Tab panel
> - `Tab` - once to navigate into the group of tabs and once to navigate out of the group of tabs
> - `↑`/`↓` or `←`/`→` - choose and activate previous/next tab.

but the autocomplete section states that arrow keys are the way to navigate into and out of the composite control.

> Autocomplete
> - Type to begin filtering
> - `↑`/`↓` - navigate to an option
> - `Enter` - select an option

But I wonder if having it be tab-able is something _our_ users would want? It's probably too early to say so maybe removing this is a better option till usability testing.

Comment (firestack) [reply assets/src/components/groupedAutocomplete.tsx]: 1242d36129eb26966573a1ae2d2b2906d3bf3519

File: assets/src/components/groupedAutocomplete.tsx
Comment (firestack) [reply assets/src/components/groupedAutocomplete.tsx:185]: I couldn't think of a way to to this outside of the reducer, but I'm open to suggestions.

This was for making it possible to refocus the input element after the cursor tries to move past the last or first option.
See https://github.com/mbta/skate/blob/da2eb59366c7669c7cdef03f6c189de76d166725/assets/src/components/searchForm.tsx#L308-L309 .

**Review by lemald — APPROVED** [review]
Okay, at this point I don't have any more specific comments and I don't have any objections to the overall approach, so I'm going to approve. That said I think it would probably be good to also wait for @KaylaBrady to approve on this one just since this is a big one and she's been catching stuff that I hadn't thought of.

**Review by KaylaBrady — APPROVED** [review]
A few minor discussion threads open, but nothing blocking!

File: assets/src/components/groupedAutocomplete.tsx
Comment (firestack) [reply assets/src/components/groupedAutocomplete.tsx:185]: Okay so I tried implementing a way I thought would work, but in the end I couldn't quite figure out how to keep the same functionality.
I've pushed up my WIP at b699eb24a8a4c4874b90fe82e4865bb5e7759f89, and I added a condition to the test to fire two keydown events in a row, instead of just one, which shows the problem I'm facing with implementing this.

Essentially, I need the cursor state to exist for one render, and I'm not sure how to go about that at this moment.
If you have any thoughts on how to approch this or where I have gone wrong I'd appreciate the insight, but right now I think I need to take a break and come back to it, and may need to leave this to be addressed in the future.

Comment (KaylaBrady) [reply assets/src/components/groupedAutocomplete.tsx:185]: Hmmm yeah I see what you mean! A simple change doesn't jump out at me, I think this is fine as-is without trying to move the side-effect.

**Pull request merged**
